Add "intercut" post-processing function

zerebubuth opened the pull request:
This is part of a potential fix for https://github.com/mapzen/vector-datasource/issues/136, and implements the algorithm for splitting up features in one layer based on features in another, and transferring attributes across based on whether the split part is inside or outside the feature in the other layer.

The code to actually use this function is split between the tilequeue and tileserver repos, so see below for the cross-links to those PRs.

zerebubuth committed:
Add no-op intercut implementation.

zerebubuth committed:
Implemented intercut with a very naive algorithm.

rmarianski commented on TileStache/Goodies/VecTiles/transform.py:
```diff
@@ -366,3 +366,90 @@ def tags_name_i18n(shape, properties, fid, zoom):
+    # which have the names of the base and cutting layer.
+    # it would seem to be better to use a dict() for
+    # layers, and this will give odd results if names are
+    # allowed to be duplicated.
```
:+1: Maybe we make this change after the pr lands?

zerebubuth replied:
I wasn't sure whether it was easy to change. Does TileStache want `feature_layers` to be a `list` too, or is it just in tilequeue? I figured that's the kind of change which could easily break other stuff that depends on the ordering of the layers in the `list` and, while we can recover that because we have an "all layer names" list, it seems like a pretty big change. I'm definitely up for coming back to this later!

rmarianski replied:
Yea, I think it might end up being an invasive change. I haven't double checked, but my gut tells me that the relevant tilestache bits that expect a certain shape from `feature_layers` are the formatters. We can either update those to expect a dict, or just transform the dict into a list right before passing to the formatters, which should be the last step in the chain.

I don't expect anything depends on the ordering currently, although maybe it would make sense to think about that. I would guess that it would match the configuration ordering of the layers in the all layer, but I'm not 100% sure that's the case.

I just created a couple of other issues related to formatters that I had in mind, namely https://github.com/mapzen/tilequeue/issues/23 and https://github.com/mapzen/tilequeue/issues/24, and it might be easier to look at all these more closely together because they touch the same parts of the code.

But yea, I was thinking of looking at all this in the future and not in this pr.

rmarianski commented on TileStache/Goodies/VecTiles/transform.py:
```diff
@@ -366,3 +366,90 @@ def tags_name_i18n(shape, properties, fid, zoom):
+    for cutting_feature in cutting_features:
+        cutting_shape, cutting_props, cutting_id = cutting_feature
+        cutting_attr = None
+        if attribute is not None and attribute in cutting_props:
```
How about doing it this way?

``` python
if attribute is not None:
    cutting_attr = cutting_props.get(attribute)
```

zerebubuth replied:
Ah, yes, that's much more readable, thanks.

rmarianski commented on TileStache/Goodies/VecTiles/transform.py:
```diff
@@ -366,3 +366,90 @@ def tags_name_i18n(shape, properties, fid, zoom):
+    # didn't find one or other layer - what's the appropriate
+    # thing to do here, raise an error?
+    if base is None or cutting is None:
+        return None
```
I think an assert should be fine here, because that suggests a config problem. Maybe something like:

``` python
assert base is not None and cutting is not None, 'intercut config error'
```

zerebubuth replied:
Cool, thanks.

rmarianski commented on TileStache/Goodies/VecTiles/transform.py:
```diff
@@ -366,3 +366,90 @@ def tags_name_i18n(shape, properties, fid, zoom):
+    # TODO: this is a very simple way of doing this, and would
+    # probably be better replaced by something that isn't O(N^2)
+    # and perhaps even unioned features with the same attribute
+    # together first.
```
Hhmm, what do you mean by unioned features?

Would other approaches include creating a spatial index in memory and using that to cut down on some of the checks?

zerebubuth replied:
What I was thinking was that we want to minimise the number of cuts, as each split road is going to have overhead associated with having its own attributes, arcs, etc... Given that all we are interested in, for the purposes of splitting, is the attribute on the cutting layer then perhaps it makes sense to union together all the features in the cutting layer which share the same attribute value. Clipping against each of those, from highest priority to lowest, and not re-clipping the parts which are inside higher priority polygons should give us the minimum number of cuts.

Starting with (please excuse the ASCII art...):

```
  aaaaaaa
  a     a
  a   bbabbbb
  a   b a   b
  a ccbcacc b
--a-c-b-a-c-b--
  aaaaaaa c b
    c b   c b
    c bbbbbbb
    c     c
    ccccccc
```

Where the features `[a, b, c]` are in the cutting layer, and `a` and `c` have the same attribute, which is lower priority than `b`'s attribute. A line going through the center `---` would be cut into 7 pieces, and we've done 7 intersection operations. If we union together `a` and `c` and cut first against `b`, then we get:

```
  aaaaaaa
  a     a
  a   bbabbbb
  a   b a   b
  a   b aaa b
--a---b=====b--
  aaa b   a b
    a b   a b
    a bbbbbbb
    a     a
    aaaaaaa
```

By first cutting against the highest priority polygon, `b`, and setting aside the inside section (marked `===`) as it can't be cut by anything higher priority, and then the union of `a` and `c`, we get 4 sections instead of 7, and we've done 2 intersection operations.

Creating spatial indexes is an additional potential optimisation, as is using shapely's "prepared geometries", and I wouldn't want to guess what the impact would be of using either of these without trying to measure it first...

rmarianski replied:
Ah ok, makes sense. Thanks for the explanation, and fwiw I digged the ascii art! :smile:  

But for the union to help, it would mean that there would need to be some overlaps in the first place. In this case, multiple landuse features would need to intersect and share the same property. I wouldn't expect that to be a common occurrence, but I don't have as good of a feel about the data. Does this sort of thing happen often (enough)? I know it does for water, and we've talked a little bit about that in https://github.com/mapzen/vector-datasource/issues/140. But for the water case, I'd assume that we'd want to merge the features first so that we wouldn't have overlaps anyway. That's probably more specific to water than the other layers though.

rmarianski commented on TileStache/Goodies/VecTiles/transform.py:
```diff
@@ -366,3 +366,90 @@ def tags_name_i18n(shape, properties, fid, zoom):
+            cutting_attr = cutting_props[attribute]
+
+        new_features = []
+        for index, base_feature in enumerate(base_features):
```
`index` isn't used later, so we can just iterate through here.

zerebubuth replied:
D'oh. Yup, that was a remnant from when I was updating `base_features` directly rather than building `new_features`.

zerebubuth committed:
Fixes for readability. Now asserts instead of returning None on config error.

zerebubuth committed:
Wrote a more sophisticated version of the intercut algorithm, using a list of indexes to speed up intersection calculations.

zerebubuth committed:
'Fix' problem where base/cutting layer isn't available.

This can happen when the client selects a subset of layers which don't include both the base and cutting layers. If the base layer isn't selected, then the appropriate thing to do, which this patch does, is skip the post-processing step.

However, if the client selects the base layer, but not the cutting layer, then it's too late to do anything about it in the post-processor. The appropriate thing may be to not perform the cutting process, or it might be to change the code so that it additionally selects the cutting layer and filters it out when the client doesn't want it.

rmarianski commented:
:+1: Feel free to merge in.

I added a comment/question: https://github.com/mapzen/TileStache/commit/6fe1347f15fa59bfa51bb2a9ec96916b0a5a2252#commitcomment-13120898

zerebubuth committed:
Made intercut attribute a required parameter, and added an assert for config error if it's not present.
